openstack-api-charm-creation-guide.md

```diff
@@ -0,0 +1,240 @@
+from charm.openstack.charm import OpenStackCharmFactory, OpenStackCharm
+from charm.openstack.adapters import OpenStackRelationAdapters
+
+class CongressCharm(OpenStackCharm):
+    # Packages the service needs installed
+    packages = ['congress-server', 'congress-common', 'python-antlr3', 'python-pymysql']
+
+    # Init services the charm manages
+    services = ['congress-server']
+
+    # Standard interface adapters class to use.
+    adapters_class = OpenStackRelationAdapters
+
+    # Ports that need exposing.
+    default_service = 'congress-api'
+    api_ports = {
+        'congress-api': {
+            PUBLIC: 1789,
+            ADMIN: 1789,
+            INTERNAL: 1789,
+        }
+    }
+
+    # Database sync command used to initalise the schema.
+    sync_cmd = ['congress-db-manage', '--config-file', '/etc/congress/congress.conf', 'upgrade', 'head']
+
+    # The restart map defines which services should be restarted when a given file changes
+    restart_map = {
+        '/etc/congress/congress.conf': ['congress-server'],
+        '/etc/congress/api-paste.ini': ['congress-server'],
+        '/etc/congress/policy.json': ['congress-server'],
+    }
+```
+
+Finally create a charm factory, this is used to allow a different Charm class to be used for different Openstack releases. In this case the same charm class will be used for all Liberty and above releases.
+
+```
+class CongressCharmFactory(OpenStackCharmFactory):
+
+    releases = {
+        'liberty': CongressCharm
+    }
+
+    first_release = 'liberty'
+```
+
+## Add Congress code to react to events
+
+### Install Congress Packages
+
+The reactive framework is going to emit events that the Congress charm can react to. the charm needs to define how its going to react to these events and also raise new events as needed.
+
+The first action a charm needs to do is to install the Congress code. This is by done running the install method from CongressCharm created earlier. Edit charm/reactive/handlers.py ...
+
+```
+import charms.reactive as reactive
+import charm.openstack.congress as congress
+import charmhelpers.contrib.openstack.utils as ch_utils
+
+charm = None
+
+def get_charm():
+    # Retrieve a charm instance from the charm factory.
+    global charm
+    if charm is None:
+        charm = congress.CongressCharmFactory.charm(
+            release=ch_utils.os_release('congress-common', base='liberty'))
+    return charm
+
+@reactive.hook('install')
+def install_packages():
+    # When install hook fires install charm packages
+    get_charm().install()
+```
+
+### Configure Congress Relation
+
+At this point the charm could be built and deployed and it would deploy a unit, and install congress. However there is no code to specify how this charm should interact with the services to depend on. For example when joining the database the charm needs to specify the user and database it requires. The following code configures the relations with the dependant services.
+
```

Comment:
This is also sort of obvious but it wouldn't hurt to explicitly say this goes in the charm/reactive/handlers.py file.